### PR TITLE
Manage exception in get_access_token

### PR DESCRIPTION
**Pull request by ollo69**
Hi,

this PR is to complete [this](https://github.com/home-assistant/core/pull/34057) Home Assistant PR.
Basically Tuya API will raise a new exception (TuyaNETException) if connection fails at startup.
This will be used in HA to retry component creation.

### CODE REVIEW

File: tuyaha/tuyaapi.py [hunk 1]
```diff
@@ -77,7 +84,6 @@ def get_access_token(self):
     def check_access_token(self):
         if SESSION.username == "" or SESSION.password == "":
             raise TuyaAPIException("can not find username or password")
-            return
```
Comment (PaulAnnekov) [inline tuyaha/tuyaapi.py:80]: Why did you remove this `return`? I see no sense.

File: tuyaha/tuyaapi.py [hunk 2]
```diff
@@ -45,16 +46,22 @@ def init(self, username, password, countryCode, bizType=""):
...
+                data={
+                    "userName": SESSION.username,
+                    "password": SESSION.password,
+                    "countryCode": SESSION.countryCode,
+                    "bizType": SESSION.bizType,
+                    "from": "tuya",
+                },
+            )
+        except RequestsConnectionError as ex:
+            if "NewConnectionError" in str(ex):
```
Comment (PaulAnnekov) [inline tuyaha/tuyaapi.py:61]: 1. It's a bad practise to check stringified error contents, because it contains human-readable description which is subject to change.
2. Actually, you don't need such specific check. I read your PR in HA and you originally wanted "distinguishing between ConnectionError for network problems or other conditions (such as an incorrect password).". So you need to handle any `ConnectionError` and `HTTPError` with 5xx statuses. This will mean exactly what you want - network problem or server errors. Here is an information about raised errors https://2.python-requests.org/en/master/user/quickstart/#errors-and-exceptions.

So your code should look like this:
```
except RequestsConnectionError as ex:
    raise TuyaNetException from ex # Not need to uppercase "NET", it's not abbreviation
except RequestsHTTPError as ex:
    if response.status_code >= 500:
        raise TuyaServerException from ex
    raise ex
```
Also intriduce `TuyaServerException` exception. After that catch `TuyaNetException` and `TuyaServerException` in HA component.

**Comment by PaulAnnekov** [comment]
Thanks for PR ;). Fix issues and I will merge it.

File: tuyaha/tuyaapi.py (hunk 1 above)
Comment (ollo69) [reply tuyaha/tuyaapi.py:80]: Because is unreachable after a "raise"

File: tuyaha/tuyaapi.py (hunk 2 above)
Comment (ollo69) [reply tuyaha/tuyaapi.py:61]: 1. I totally agree with you, but I didn't know how to manage :)
2. Are you sure that we will receive a "RequestsHTTPError"? Because in HA I was always receiving a RequestsConnectionError.

File: tuyaha/tuyaapi.py (hunk 1 above)
Comment (PaulAnnekov) [reply tuyaha/tuyaapi.py:80]: Ah, me stupid)))). You're right.

File: tuyaha/tuyaapi.py (hunk 2 above)
Comment (PaulAnnekov) [reply tuyaha/tuyaapi.py:61]: I'm sure. ConnectionError means we even didn't get any reply from server. Only reply can say that there is something wrong with your credentials.
Just checked on my HA instance. Made username wrong and got:
```
home_assistant    | 2020-04-26 15:46:37 ERROR (MainThread) [homeassistant.setup] Error during setup of component tuya
home_assistant    | Traceback (most recent call last):
home_assistant    |   File "/usr/src/homeassistant/homeassistant/setup.py", line 175, in _async_setup_component
home_assistant    |     component.setup, hass, processed_config  # type: ignore
home_assistant    |   File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run
home_assistant    |     result = self.fn(*self.args, **self.kwargs)
home_assistant    |   File "/usr/src/homeassistant/homeassistant/components/tuya/__init__.py", line 63, in setup
home_assistant    |     tuya.init(username, password, country_code, platform)
home_assistant    |   File "/usr/local/lib/python3.7/site-packages/tuyaha/tuyaapi.py", line 43, in init
home_assistant    |     self.get_access_token()
home_assistant    |   File "/usr/local/lib/python3.7/site-packages/tuyaha/tuyaapi.py", line 65, in get_access_token
home_assistant    |     raise TuyaAPIException(message)
home_assistant    | tuyaha.tuyaapi.TuyaAPIException: Get accesstoken failed. Username or password error!

```

Comment (ollo69) [reply tuyaha/tuyaapi.py:61]: You are right, anyway I don't think that catching "RequestsHTTPError" make sense because we will not have this error until you call `response.raise_for_status()`.
You are already managing HTTP error with your check on response content, alternative solution should be this if you like to add TuyaServerException  but I don't think it make sense:

```
        except RequestsConnectionError as ex:
            raise TuyaNetException from ex

        try:
            response.raise_for_status()
        except RequestsHTTPError as ex:
            if response.status_code >= 500:
                raise TuyaServerException from ex
            raise ex
```

My problem is that applying this fix will cause HA test to fail, please check [here](https://github.com/home-assistant/core/pull/34057/checks?check_run_id=619834272).
This probably happen because HA test raise generic ConnectionError to simulate connection problem and is not able to understand the difference, maybe raising the new exception from the library will fix this problem in test?
 
What do you think?

Comment (PaulAnnekov) [reply tuyaha/tuyaapi.py]: > we will not have this error until you call response.raise_for_status().

You're right.

> You are already managing HTTP error with your check on response content

When service responds with 5xx error it can send empty body and `response_json = response.json()` line will emit some generic exception. It's up to you whether you want to handle this case in this PR.

> My problem is that applying this fix will cause HA test to fail

Try to re-run that build. Its error logs says the problem is with tplink component, not tuya one. Looks like unstable tplink test.
IMHO, it's a good practise to have own exceptions in a library. Because library user should not care what are you using inside to make HTTP requests (e.g. `request` lib).

Comment (ollo69) [reply tuyaha/tuyaapi.py]: Fine, take a look to last Commit and let me know if looks fine for you.

**Comment by PaulAnnekov** [comment]
@ollo69 thanks, merged :+1: .

**Comment by ollo69** [comment]
Thanks!!!
